### PR TITLE
Adaptive stepsize

### PR DESCRIPTION
**Pull request by RemDelaporteMathurin**
## Proposed changes

This is an attempt at adding adaptive timestepping.

I've taken a [similar approach to MOOSE](https://mooseframework.inl.gov/source/timesteppers/IterationAdaptiveDT.html) where the users give a target number of iterations and a growth factor and a cutback factor.

If the number of newton iterations is larger than the target number, the stepsize is reduced (multiplied by cutback factor). Inversely when it's lower than the target, the stepsize is increased (growth factor).

@jhdark I've noticed we have the concept of the stepsize living in different places.
First we have `HTransportProblem.settings.stepsize` AND `HTransportProblem.dt` (which is just the `fenicsx.Constant` object).
I feel like we should remove this redundancy to improve clarity. Doesn't have to be in this PR but I felt like we could kick off the discussion now

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: festim/stepsize.py [hunk 1]
```diff
@@ -10,19 +10,74 @@ class Stepsize:
 
     Attributes:
         initial_value (float, int): initial stepsize.
+        growth_factor (float): factor by which the stepsize is
+            increased when adapting
+        cutback_factor (float): factor by which the stepsize is
+            decreased when adapting
+        target_nb_iterations (int): number of Newton iterations
+            over (resp. under) which the stepsize is increased
+            (resp. decreased)
```
Comment (jhdark) [inline festim/stepsize.py:19]: Should these not be optional arguments too?

File: festim/stepsize.py [hunk 2]
```diff
@@ -10,19 +10,74 @@ class Stepsize:
...
         self.initial_value = initial_value
+        self.growth_factor = None
+        self.cutback_factor = None
+        self.target_nb_iterations = None
+
+        # TODO should this class hold the dt object used in the formulation
+
+    @property
+    def adaptive(self):
+        return self.growth_factor or self.cutback_factor or self.target_nb_iterations
```
Comment (jhdark) [inline festim/stepsize.py:35]: This should also be in the documentation

File: festim/stepsize.py [hunk 3]
```diff
@@ -10,19 +10,74 @@ class Stepsize:
...
+    @cutback_factor.setter
+    def cutback_factor(self, value):
+        if value is not None:
+            if value > 1:
+                raise ValueError("cutback factor should be smaller than one")
+
+        self._cutback_factor = value
+
+    def modify_value(self, value, nb_iterations, t=None):
+        if not self.is_adapt(t):
```
Comment (jhdark) [inline festim/stepsize.py:62]: If you already have a property adaptive, why not use it here? As it only needs to be modified if the stepsize is adaptive 
```suggestion
        if not self.adaptive:
```

**Review by jhdark — CHANGES_REQUESTED** [review]
Think we can change a few small things here, but otherwise, its looking good and will be incredibly useful

File: festim/stepsize.py (hunk 3 above)
Comment (RemDelaporteMathurin) [reply festim/stepsize.py:62]: Right I understand the confusion. I guess I forgot to show why/how we use `is_adapt(t)`

The idea is that sometimes you want to adapt the timestep at certain times only.

By default, it's always. But say we want to adapt if for t between 0, 100, and 200-1000, etc. then you can override the `is_adapt` method to reflect this behaviour

File: festim/stepsize.py [hunk 4]
```diff
@@ -10,19 +10,74 @@ class Stepsize:
...
         Args:
-            mesh (dolfinx.mesh.Mesh): the domain mesh
+            t (float): the current time
+
         Returns:
-            fem.Constant: the dt value
+            bool: True if needs to adapt, False otherwise.
         """
-        return F.as_fenics_constant(self.initial_value, mesh)
+        return True
```
[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply festim/stepsize.py:83]: See reply above

File: festim/stepsize.py [hunk 5]
```diff
@@ -10,19 +10,74 @@ class Stepsize:
...
     def __init__(
         self,
         initial_value,
     ) -> None:
         self.initial_value = initial_value
+        self.growth_factor = None
+        self.cutback_factor = None
+        self.target_nb_iterations = None
+
+        # TODO should this class hold the dt object used in the formulation
```
[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply festim/stepsize.py:31]: If you don't mind I'd rather merge this and then focus on it on another PR

File: festim/stepsize.py [hunk 6]
```diff
@@ -7,22 +7,89 @@ class Stepsize:
 
     Args:
         initial_value (float, int): initial stepsize.
+        grwoth_factor (float, optional): factor by which the stepsize is
```
Comment (jhdark) [inline festim/stepsize.py:10]:
```suggestion
        growth_factor (float, optional): factor by which the stepsize is
```

**Review by jhdark — APPROVED** [review]
Looks good

**Pull request merged**